keps/sig-api-machinery/2309-idl/README.md

```diff
@@ -0,0 +1,636 @@
+   can do code generation, etc from their native environments
+
+3. **Be a good code generation input**: it should be relatively
+   straightforward to generate idiomatic types in Go.  While it’s
+   a non-goal to actually implement output for other languages, it should
+   be possible to do so and get idiomatic types (consider this "external
+   work" from goal 5).
+
+4. **Be a good OpenAPI generation input**: it should be relatively
+   straightforward to compile our IDL into Kubernetes OpenAPI.
+
+5. **Enable future/external work & experimentation**: there are many
+   possible follow-up improvements or experiments that can be done with an
+   IDL. While implementing these explicitly does not fall under the
+   concrete success criteria for this proposal, the IDL pipeline should be
+   flexible enough to support developing them as future work by us or
+   external tooling by third-party projects.
+
+## Non-Goals
+
+1. **Be a custom serialization format**: unlike proto, we’re not
+   interested in *directly* describing a serialization format.  Instead,
+   we’re interested in being able to output types.go, proto IDL, code
+   suitable for serializing in JSON, etc
+
+2. **Support all features of OpenAPI**: the Kubernetes API guidelines
+   suggest avoiding certain OpenAPI constructs, while the structural
+   schema requirements explicitly reject others.  We’re interested in an
+   IDL that guides people down the right path, and that doesn’t support
+   constructs that aren’t allowed
+
+3. **Deprecate/replace OpenAPI everywhere**: it’s not currently intended
+   that this replace all uses of OpenAPI -- OpenAPI is still part of the
+   CRD specification, will still be published by the API server, etc.
+
+4. **Change k/k’s Go type representation**: the existing codebase is
+   large and intricate.  For the sake of compatibility, initial efforts
+   will focus around generating 1-1 compatible Go with the code that
+   exists currently.
+
+## Proposal
+
+***Author's Note***: *Please* read [Alternatives](#alternatives) before
+commenting "why not" or "have you considered" here.
+
+### Syntax
+
+See [the attached proposal][syntax] for grammar, syntax design, and
+in-depth examples.
+
+[syntax]: ./syntax.md
+
+### Output Scope
+
+The intial scope of this proposal is to generate two concrete artifacts:
+**CRDs** (i.e. what we have in controller-tools today) and **Go types**
+(i.e. generate types.go).
+
+The format is intended to be flexible enough that we can eventually port
+all of the generators in code-generator to make use of the IDL.  These are
+considered follow-up work after the MVP.
+
+### Concrete KubeBuilder & k/k Changes
+
+* **All types in k/k are rewritten to use KDL**.  All Go types in k/k are
+  generated.
+   
+  This guides core k/k towards KDL usage, and enables third-parties to
+  interact with k/k types well from KDL.
+
+* **The KubeBuilder project is updated to support KDL, and the pseudo-IDL
```

Comment:
So generator style tools would use the IDL.

Is there something that prevents this goal from succeeding if k/k (for some reason) is too expensive / complex to fully support?

Reply:
I've added a section on this now: https://github.com/DirectXMan12/enhancements/blob/kep/idl/keps/sig-api-machinery/2309-idl/README.md#addendum-why-do-we-need-to-do-this-in-kk